diff --git a/server/NatController.cpp b/server/NatController.cpp
index cda8f5f..cb2c16a 100644
--- a/server/NatController.cpp
+++ b/server/NatController.cpp
@@ -376,7 +376,7 @@ int NatController::setForwardRules(bool add, const char *intIface, const char *e
         goto err_return;
     }
 
-    if (runCmd(ARRAY_SIZE(cmd4), cmd4) && add) {
+    if (runCmd(ARRAY_SIZE(cmd4), cmd4) && add && false /* STOPSHIP */) {
         rc = -1;
         goto err_rpfilter;
     }
